_posts/2020-01-24-numpy-broadcasting/numpy-broadcasting.Rmd

```diff
@@ -0,0 +1,640 @@
+
+First check -- `(4, 1)` added to `(4,)` should yield `(4, 4)`:
+
+```{r}
+a <- tf$ones(shape = c(4L, 1L))
+a
+# tf.Tensor(
+# [[1.]
+#  [1.]
+#  [1.]
+#  [1.]], shape=(4, 1), dtype=float32)
+
+b <- tf$constant(c(1, 2, 3, 4))
+b
+# tf.Tensor([1. 2. 3. 4.], shape=(4,), dtype=float32)
+
+a + b
+# tf.Tensor(
+# [[2. 3. 4. 5.]
+# [2. 3. 4. 5.]
+# [2. 3. 4. 5.]
+# [2. 3. 4. 5.]], shape=(4, 4), dtype=float32)
+```
+
+And second, when we add tensors with shapes `(3, 3)` and `(3,)`, the 1-d tensor should get added to every row (not every column):
+
+
+```{r}
+a <- tf$constant(matrix(1:9, ncol = 3, byrow = TRUE), dtype = tf$float32)
+a
+# tf.Tensor(
+# [[1. 2. 3.]
+#  [4. 5. 6.]
+#  [7. 8. 9.]], shape=(3, 3), dtype=float32)
+
+b <- tf$constant(c(100, 200, 300))
+b
+# tf.Tensor([100. 200. 300.], shape=(3,), dtype=float32)
+
+a + b
+# tf.Tensor(
+# [[101. 202. 303.]
+#  [104. 205. 306.]
+#  [107. 208. 309.]], shape=(3, 3), dtype=float32)
+```
+
+Now back to the initial `matmul` example. 
+
+## Back to the puzzles
+
+The [documentation for matmul says](https://www.tensorflow.org/api_docs/python/tf/linalg/matmul),
+
+> The inputs must, following any transpositions, be tensors of rank >= 2 where the inner 2 dimensions specify valid matrix multiplication dimensions, and any further outer dimensions specify matching batch size.
+
+So here (see code just below), the inner two dimensions look good -- `(2, 3)` and `(3, 2)` -- while the one (one and only, in this case) batch dimension shows mismatching values `2` and `1`, respectively.
+A case for broadcasting thus: Both "batches" of `a` get matrix-multiplied with `b`.
+
+```{r}
+a <- tf$constant(keras::array_reshape(1:12, dim = c(2, 2, 3)))
+a 
+# tf.Tensor(
+# [[[ 1.  2.  3.]
+#   [ 4.  5.  6.]]
+# 
+#  [[ 7.  8.  9.]
+#   [10. 11. 12.]]], shape=(2, 2, 3), dtype=float64)
+
+b <- tf$constant(keras::array_reshape(101:106, dim = c(1, 3, 2)))
+b  
+# tf.Tensor(
+# [[[101. 102.]
+#   [103. 104.]
+#   [105. 106.]]], shape=(1, 3, 2), dtype=float64)
+
+c <- tf$matmul(a, b)
+c
+# tf.Tensor(
+# [[[ 622.  628.]
+#   [1549. 1564.]]
+# 
+#  [[2476. 2500.]
+#   [3403. 3436.]]], shape=(2, 2, 2), dtype=float64) 
+```
+
+
+Let's quickly check this really is what happens, by multiplying both batches separately:
+
+```{r}
+
+tf$matmul(a[1, , ], b)
+# tf.Tensor(
+# [[[ 622.  628.]
+#   [1549. 1564.]]], shape=(1, 2, 2), dtype=float64)
+
+tf$matmul(a[2, , ], b)
+# tf.Tensor(
+# [[[2476. 2500.]
+#   [3403. 3436.]]], shape=(1, 2, 2), dtype=float64)
+```
+
+
+Is it too weird to be wondering if broadcasting would also happen for matrix dimensions? E.g., could we try `matmul`ing tensors of shapes `(2, 4, 1)` and `(2, 3, 1)`, where the `4 x 1` matrix would be broadcast to `4 x 3`? -- A quick test shows that no.
+
+To see how really, when dealing with TensorFlow operations, it pays off overcoming one's inner ADHD and actually consult the documentation, let's try another one.
```

Comment:
I try to avoid using diagnoses as slang, maybe inner desire? (Don't feel bad—lots of people do it; I'm probably an outlier in avoiding it).

Reply:
Hm... I think you mean it's cheap, kinda?
Of course when you say "lots of people do it" I have no choice than change (ehem)
:-)))

Reply:
-> "initial reluctance"